File: 145.cpp
#include<iostream>
#include<stack>
#include "Tree.h"

using namespace std;

class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        postorder(root,res);
        return res;
    }
    void postorder(TreeNode* root,vector<int>& res){
        if(root){
            postorder(root->left,res);
            postorder(root->right,res);
            res.push_back(root->val);
            
        }
    }
};

class Solution2 {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> res;
        stack<TreeNode*> s;
        TreeNode* curNode=root;
        TreeNode* temp;
        while(curNode||!s.empty()){
            if(curNode){
                s.push(curNode);
                curNode=curNode->left;
            }else{
                TreeNode* temp=s.top();
                if(temp->right){
                    curNode=temp->right;
                    temp->right=nullptr;
                }else{
                    res.push_back(temp->val);
                    s.pop();
                }
            }
        }
        return res;
    }
};

File: Python/145.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        stack = []
        
        while stack or root:
            while root:
                stack.append(root)
                root = root.left
                
            root = stack.pop()
            tmp = root.right
            if not root.right:
                res.append(root.val)
            else:
                root.right = None
                stack.append(root)
            root = tmp
        return res


class Solution(object):
    def postorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        res = []
        self.dfs(root, res)
        return res

    def dfs(self, root, res):
        if root:
            self.dfs(root.left, res)
            self.dfs(root.right, res)
            res.append(root.val)
